sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryStatusListener.scala

```diff
@@ -18,104 +18,161 @@
   override def onQueryProgress(event: StreamingQueryListener.QueryProgressEvent): Unit = {
-    val batchTimestamp = parseProgressTimestamp(event.progress.timestamp)
-    val queryStatus = activeQueryStatus.getOrDefault(
-      event.progress.runId,
-      new StreamingQueryUIData(event.progress.name, event.progress.id, event.progress.runId,
-        batchTimestamp))
-    queryStatus.updateProcess(event.progress, streamingProgressRetention)
+    val runId = event.progress.runId
+    val batchId = event.progress.batchId
+    val timestamp = event.progress.timestamp
+    val querySummary = store.read(classOf[StreamingQueryUIData], runId)
+    val progressIdQueue =
+      querySummary.progressIdQueue ++ Seq(getUniqueId(runId, batchId, timestamp))
+    store.write(new StreamingQueryProgressWrapper(event.progress))
+    while(progressIdQueue.length >= streamingProgressRetention) {
+      val uniqueId = progressIdQueue.dequeue
+      store.delete(classOf[StreamingQueryProgressWrapper], uniqueId)
+    }
+    store.delete(classOf[StreamingQueryUIData], runId)
+    store.write(new StreamingQueryUIData(
+      querySummary.name,
+      querySummary.id,
+      querySummary.runId,
+      progressIdQueue,
+      querySummary.startTimestamp,
+      querySummary.isActive,
+      querySummary.exception
+    ))
   }
 
   override def onQueryTerminated(
       event: StreamingQueryListener.QueryTerminatedEvent): Unit = synchronized {
-    val queryStatus = activeQueryStatus.remove(event.runId)
-    if (queryStatus != null) {
-      queryStatus.queryTerminated(event)
-      inactiveQueryStatus += queryStatus
-      while (inactiveQueryStatus.length >= inactiveQueryStatusRetention) {
-        inactiveQueryStatus.dequeue()
-      }
-    }
-  }
-
-  def allQueryStatus: Seq[StreamingQueryUIData] = synchronized {
-    activeQueryStatus.values().asScala.toSeq ++ inactiveQueryStatus
+    val querySummary = store.read(classOf[StreamingQueryUIData], event.runId)
+    store.delete(classOf[StreamingQueryUIData], event.runId)
+    store.write(new StreamingQueryUIData(
+      querySummary.name,
+      querySummary.id,
+      querySummary.runId,
+      querySummary.progressIdQueue,
+      querySummary.startTimestamp,
+      false,
+      querySummary.exception
+    ))
   }
 }
 
 /**
  * This class contains all message related to UI display, each instance corresponds to a single
  * [[org.apache.spark.sql.streaming.StreamingQuery]].
  */
-private[ui] class StreamingQueryUIData(
+private[sql] class StreamingQueryUIData(
     val name: String,
     val id: UUID,
-    val runId: UUID,
-    val startTimestamp: Long) {
+    @KVIndexParam val runId: UUID,
+    val progressIdQueue: Queue[String],
+    val startTimestamp: Long,
+    val isActive: Boolean,
+    val exception: Option[String]) {
 
-  /** Holds the most recent query progress updates. */
-  private val progressBuffer = new mutable.Queue[StreamingQueryProgress]()
+  private var storeOption: Option[ElementTrackingStore] = None
```

Comment:
Can we get rid of saving KVStore in UIData? I think all the store related operation should be placed in `StreamingQueryStatesStore`?

Reply:
OK, let me materialize UI data directly instead of putting `KVStore` into `StreamingQueryUIData`.